### PR TITLE
Fix to optimize getting table index structure for large databases

### PR DESCRIPTION
**Pull request by iliajie**
Hey Jamie!

This PR will optimize getting table index structure on the huge databases.

It was tested with TikiWiki database which over 400 tables and indexes.

The performace boost is around 500%.

### CODE REVIEW

**Comment by iliajie** [comment]
@marclaporte Thanks for reporting this issue in the first place.

If you have Webmin 2.200+ installed, you can apply this patch to run tests as simply as:

```
webmin patch https://github.com/webmin/webmin/commit/b03f1
```

To be clear, for me, the load time dropped from 9.5 seconds to 1.5 seconds.

File: mysql/mysql-lib.pl [hunk 1]
```diff
@@ -1304,6 +1304,37 @@ sub mysql_size_input
 		    [ "M", "MB" ], [ "G", "GB" ] ]);
 }
 
+# get_table_index_stats(db)
+# Retrieves index stats for all tables in the given database
+sub get_table_index_stats
+{
+my ($db) = @_;
+my $table_list = join(", ", map { "'$_'" } &list_tables($db));
```
Comment (jcameron) [inline mysql/mysql-lib.pl:1312]: There's a quoting function in mysql-lib.pl you can use instead of doing it yourself

File: mysql/mysql-lib.pl [hunk 2]
```diff
@@ -1304,6 +1304,37 @@ sub mysql_size_input
...
+        SUB_PART,
+        PACKED,
+        NULLABLE,
+        INDEX_TYPE,
+        COMMENT,
+        INDEX_COMMENT
+    FROM 
+        INFORMATION_SCHEMA.STATISTICS
+    WHERE 
+        TABLE_SCHEMA = '$db' AND TABLE_NAME IN ($table_list);
```
Comment (jcameron) [inline mysql/mysql-lib.pl:1332]: Use a proper SQL parameter instead of substituting $db into the string

Comment (iliajie) [reply mysql/mysql-lib.pl]: Yes, sure! Check this patch https://github.com/webmin/webmin/pull/2234/commits/5210f623f3f1a0b8540fa6ac0d2c94689a4f1f8f.

File: mysql/mysql-lib.pl (hunk 1 above)
Comment (iliajie) [reply mysql/mysql-lib.pl]: I don't think we even need this with formal params ..

**Comment by marclaporte** [comment]
> huge databases.
> 
> It was tested with TikiWiki database which over 400 tables and indexes.

Yes, Tiki is good to push limits :-)

All Tiki's features are built-in, so it's big on a fresh install (mostly empty tables) but the number of tables doesn't grow after. It's quite the opposite model to other plugin/module centric systems like WordPress/Drupal/Joomla!/...
- https://tiki.org/FLOSS-Web-Application-with-the-most-built-in-features
- https://pluginproblems.com/

**Comment by iliajie** [comment]
Marc, did you manage to apply the patch and see how much faster it gets for you?

**Comment by marclaporte** [comment]
@iliajie I didn't try yet, but I am sure it's going to be great now!

File: mysql/mysql-lib.pl (hunk 1 above)
Comment (jcameron) [reply mysql/mysql-lib.pl]: Can a table name contain ' ?

Comment (jcameron) [reply mysql/mysql-lib.pl]: Oh never mind I see you fixed this

File: mysql/mysql-lib.pl (hunk 2 above)
Comment (jcameron) [reply mysql/mysql-lib.pl]: Excellent thanks!

**Pull request merged**